build.gradle.kts

```diff
@@ -0,0 +1,156 @@
+    group = "com.amazonaws"
+
+    repositories {
+        mavenCentral()
+        mavenLocal()
+    }
+
+    // Configure the shadow jar task, which does shading, to run after Gradle runs the jar task
+    pluginManager.withPlugin("com.github.johnrengelman.shadow") {
+        tasks {
+            named<DefaultTask>("assemble") {
+                dependsOn(named("shadowJar"))
+            }
+
+            named<ShadowJar>("shadowJar") {
+                // Suppress the "-all" suffix on the jar name, simply replace the default built jar instead
+                archiveClassifier.set("")
+            }
+
+            // Disable the conventional JAR task, and only run shadowJar instead
+            named<Jar>("jar") {
+                enabled = false
+                dependsOn(named("shadowJar"))
+            }
+        }
+    }
+
+    plugins.withId("java") {
+        configure<JavaPluginExtension> {
+            sourceCompatibility = JavaVersion.VERSION_1_8
+            targetCompatibility = JavaVersion.VERSION_1_8
+
+            withJavadocJar()
+            withSourcesJar()
+        }
+
+        dependencies {
+            // BOMs for common projects
+            add("implementation", platform("com.amazonaws:aws-xray-recorder-sdk-bom:2.7.1"))
+            add("implementation", platform("software.amazon.disco:disco-toolkit-bom:${discoVersion}"))
+            add("implementation", platform("com.fasterxml.jackson:jackson-bom:2.11.0"))
+            add("implementation", platform("com.amazonaws:aws-java-sdk-bom:1.11.837"))
+            add("implementation", platform("software.amazon.awssdk:bom:2.13.70"))
+
+            // TODO: Add build step for running Null checker
+            add("compileOnly", "org.checkerframework:checker-qual:3.4.1")
+
+            // Common test dependencies
+            add("testImplementation", "junit:junit:4.12")
+            add("testImplementation", "org.assertj:assertj-core:3.16.1")
+            add("testImplementation", "org.mockito:mockito-core:2.28.2")
+        }
+    }
+
+    plugins.withId("maven-publish") {
+        plugins.apply("signing")
+
+        // Disable publishing a bunch of unnecessary Gradle metadata files
+        tasks.withType<GenerateModuleMetadata> {
+            enabled = false
+        }
+
+        // Defer maven publish until the assemble task has finished, giving time for shadowJar to complete if present
+        tasks.withType<AbstractPublishToMaven> {
+            dependsOn(tasks.named<DefaultTask>("assemble"))
+        }
+
+        configure<PublishingExtension> {
+            publications {
+                register<MavenPublication>("maven") {
+
+                    // If the shadow plugin is present, we should publish the shaded artifact
+                    // Otherwise, just publish the standard JAR
+                    plugins.withId("java") {
+                        afterEvaluate {
+                            if (plugins.hasPlugin("com.github.johnrengelman.shadow")) {
```

Comment:
Ah I think since we replace the jar (disable default task, use the empty classifier), actually `components["java"]` picks up the shadow jar. Appears to be working here

https://github.com/anuraaga/aws-opentelemetry-java-instrumentation/blob/master/build.gradle.kts#L111

Feels a bit hacky admittedly, but it makes sure the doc / source jars are published too which I think is requirement of maven central. With this approach we need to make sure to add those

Reply:
Oh nice that worked for me too! Added the simplification.